File: network.js
// this is the network script that we load where we will register the callback
// for the networking events
let enet_states = [
        "disconnected",
        "ack-connect",
        "connection-pending",
        "connection-succeeded",
        "connected",
        "disconnect-later",
        "disconnecting",
        "ack-disconnect",
        "zombie",
];

class ServerState {
    Unconnected = 0;
    Pregame     = 1;
    Loading     = 2;
    InGame      = 3;
    PostGame    = 4;
};

var currentState = ServerState.Unconnected;

//var host = network.serve({address: "127.0.0.1", port: 20595});

function logPeer(peer) {
    let message = `peer
        state   ${enet_states[peer.state]}
        ID      ${peer.id}
        address ${peer.address}
        session ${peer.session}`;
    console.log(message);
}

network.server = (host, event) => {
  switch (event.type) {
  case "connect":
    //host.sendPacket(`+${event.peer.id}`);
    break;
  case "disconnect":
    //host.sendPacket(`-${event.peer.id}`);
    break;
  }
  logPeer(event.peer);
}

network.client = (host, event) => {
  switch (event.type) {
  case "connect":
    logPeer(event.peer);
    break;
  case "disconnect":
    break;
  case "receive":
    if (event.packet.data !== undefined) {
      console.log(event.packet.data);
    }
    break;
  }
}

//currentState = ServerState.Pregame;

File: network.cpp
void Server(ENetHost *host,  JSContext *ctx) {
  ENetEvent event;
  std::vector<ENetPeer*> peers;

  while (true) {
    while (enet_host_service(host, &event, 1000) > 0) {
      switch (event.type) {
      case ENET_EVENT_TYPE_CONNECT:
	printf("A new client connected from %x:%u.\n",
		 event.peer->address.host,
		 event.peer->address.port);
	peers.push_back(event.peer);
	break;
      case ENET_EVENT_TYPE_RECEIVE:
	printf ("A packet of length %u containing %s was received from %s on channel %u.\n",
		event.packet -> dataLength,
		event.packet -> data,
		event.peer -> data,
		event.channelID);
	enet_packet_destroy(event.packet);
	break;
      case ENET_EVENT_TYPE_DISCONNECT:
	printf("peer disconnected\n");

	//TODO(edoput) what fields of the ENetPeer structure can be used
	// to tell apart one peer from the other?
	break;
      }
    }

    ENetPacket * packet = enet_packet_create ("packet", strlen ("packet") + 1, ENET_PACKET_FLAG_RELIABLE);
    enet_host_broadcast(host, 0, packet);
  }
}

void Client(ENetHost *client,  JSContext *ctx) {
  ENetPeer* peer = Connect(client); // connection to remote

  ENetEvent event;

  while (true) {
    while (enet_host_service(client, &event, 1000) > 0) {
      switch (event.type) {
      case ENET_EVENT_TYPE_CONNECT:
	std::cout << "client: connected\n";
        break;
      case ENET_EVENT_TYPE_RECEIVE:
	std::cout << "client: receive\n";
	break;
      case ENET_EVENT_TYPE_DISCONNECT:
	std::cout << "client: disconnected\n";

	break;
      }
    }
  }
}
